coalib/bearlib/aspects/base.py

```diff
@@ -1,8 +1,94 @@
+import re
+
 from coalib.bearlib.languages import Language
 
+import coalib.bearlib.aspects
 from .taste import TasteError
 
 
+class SubaspectGetter:
+    """
+    Special "getter" class to implement get() function in aspectbase that
+    could be accessed from the aspectclass or aspectclass instance.
+    """
+
+    def __get__(self, obj, owner):
+        if obj is not None:
+            parent = obj
+            parent_qualname = type(obj).__qualname__
+        else:
+            parent = owner
+            parent_qualname = owner.__qualname__
+
+        def get_subaspect(subaspect):
+            """
+            Get a subaspect from an aspectclass or aspectclass instance.
+
+            >>> import coalib.bearlib.aspects as aspects
+            >>> metadata = aspects['Metadata']
+            >>> commit_msg = aspects['CommitMessage']
+            >>> shortlog = aspects['Shortlog']
+
+            We can get direct children.
+
+            >>> metadata.get(commit_msg)
+            <aspectclass 'Root.Metadata.CommitMessage'>
+
+            Or even a grandchildren.
+
+            >>> metadata.get(shortlog)
+            <aspectclass 'Root.Metadata.CommitMessage.Shortlog'>
+
+            Trying to access children of aspect instance will raise
+            NotImplementedError because of instanced aspect doesn't instance
+            its children and will cause wrong result.
+            See https://github.com/coala/coala/issues/4388
+
+            >>> metadata('Python').get(commit_msg)
+            Traceback (most recent call last):
+            ...
+            NotImplementedError: Cannot access children of aspect instance.
+
+            But, passing subaspect instance as argument is prohibited, because
+            it doesn't really make sense.
+
+            >>> metadata('Python').get(commit_msg('Java'))
+            Traceback (most recent call last):
+            ...
+            AttributeError: Cannot search an aspect instance using another
+            aspect instance as argument.
+
+            :param subaspect: An subaspect that we want to find in an
+                              aspectclass.
+            :return:          An aspectclass. Return None if not found.
+            """
+            # Avoid circular import
+            from .meta import issubaspect
+            if isinstance(subaspect, str):
+                subaspect = coalib.bearlib.aspects[subaspect]
+            if not issubaspect(subaspect, parent):
+                return None
+            if isinstance(subaspect, aspectbase):
```

Comment:
`if not isclass(subaspect):` is enough here since we know that we have an aspect. And it looks more intuitive and understandable. The same applies to check below